### PR TITLE
[OSD-25046]add egress verfiy for zero egress HCP cluster

### PR DESCRIPTION
**Pull request by bmeng**
## What does this PR do? / Related Issues / Jira

[OSD-25046](https://issues.redhat.com//browse/OSD-25046)
Add the new platform type for HCP Zeroegress

During the zero egress cluster creation, a list of ECR registry will be passed for validation. With the current structure, it is not easy to build the testing urls from the yaml file with single ENV matches to multiple values.

So I just simply convert the given string with to the expected one.

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have tested the functionality against gcp / aws, it doesn't cause any regression
- [ ] I have added execution results to the PR's readme

## Reviewer's Checklist

- [ ] (This needs to be done after technical review) I've run the branch on my local, verified that the functionality is ok

### CODE REVIEW

**Comment by abyrne55** [comment]
Thanks for your contribution @bmeng! This seems like a great start, but unfortunately it conflicts substantially with another rather-large PR (#268) that's in-flight. I'm going to hold this PR for now, and after #268 merges, let's reevaluate our options and determine if a more flexible approach (e.g., filling in arbitrary variables in URLs, instead of creating special variables just for ECR IDs) is appropriate.

/hold

**Comment by bmeng** [comment]
Thanks for the review @abyrne55 may I know the target date for the PR #268 (asking since it has the label `draft`)? as the change for supporting ECR and the zero egress is part of the epic which we are going to GA. I would like to get the feature as soon as possible so that we won't block the epic.

**Comment by abyrne55** [comment]
@bmeng #268 has merged :) . Note that I'll be on PTO next week. If you have any questions during that time, please post in Aurora's team channel. If you need verifier PRs approved, please ping @rafael-azevedo or @fahlmant

File: pkg/data/egress_lists/egress_lists.go
Comment (xiaoyu74) [inline pkg/data/egress_lists/egress_lists.go:117]: Assuming this function has the core logic to handle the input of the constructed ECR URL.  So, if we have 2 ECR IDs, the constructed ECR URL should be something like `"123456789,987654321.dkr.ecr.us-west-2.amazonaws.com"`  (As I can see the ECR ID is defined as 12 digits?)

I just created a test.go file to call your function `ConvertECRRegistryURLs`  passing this dummy URL string:
```
func main() {
	// Example input for multiple ECR registry IDs
	result := ConvertECRRegistryURLs("123456789012,210987654321.dkr.ecr.us-west-2.amazonaws.com")
	fmt.Println("Constructed URLs:", result)
}
```
Looks the output is concatenated as below, is that expected? 
```
go run test.go
Constructed URLs: 123456789012.dkr.ecr.us-west-2.amazonaws.com210987654321.dkr.ecr.us-west-2.amazonaws.com
```

Comment (bmeng) [reply pkg/data/egress_lists/egress_lists.go:117]: thanks for testing @xiaoyu74 
As the function is nested in the for loop within the `EgressListToString`, so it will have the prefix `https://`, suffix `:443` appended from the yaml template and and an extra space added between each url

File: pkg/data/egress_lists/egress_lists.go
Comment (rafael-azevedo) [inline pkg/data/egress_lists/egress_lists.go:75]: is this what you were using during testing?

File: pkg/data/egress_lists/egress_lists.go
Comment (rafael-azevedo) [inline pkg/data/egress_lists/egress_lists.go:116]: Why is this necessary here?Shouldn't it be one URL per input that is looped on line 97?

File: pkg/data/egress_lists/egress_lists.go
Comment (bmeng) [reply pkg/data/egress_lists/egress_lists.go:75]: ah, yes, this is for testing... I thought I should have updated back, not sure if it was missing during rebase

File: pkg/data/egress_lists/egress_lists.go
Comment (bmeng) [reply pkg/data/egress_lists/egress_lists.go:116]: in all the existing templates, it converts the URL based on the given paraemeters, like AWS_REGION, and do a one to one converting, means each item in the template will be converted to one testing URL

But for the ECR registry, one requirement is the ECR registry can be multiple. And it may not be a fixed length. So we have to do with the one item from the template to multiple URLs converting.

The easiest way I found is just convert the passed ECR slice to string and connect with `,` and then call the new function `ConvertECRRegistryURLs` to conver the joint string to separated URLs

**Comment by bmeng** [comment]
@rafael-azevedo just made a last minute change... after discuss with OCM team, it should be ok to accept a single string, rather than a slice for the ECR registry.
So it would make the implementation much more simpler

File: pkg/data/egress_lists/egress_lists.go
Comment (dustman9000) [inline pkg/data/egress_lists/egress_lists.go:67]: Where are we validating the response is a private IP?

Comment (bmeng) [reply pkg/data/egress_lists/egress_lists.go:67]: From my understanding, the network-verifier will check if the required endpoints are accessible from the given cluster vpc/subnet.
I am not sure if it is possible to ensure the network is private.

@rafael-azevedo @fahlmant could you help confirm?

Comment (dustman9000) [reply pkg/data/egress_lists/egress_lists.go:67]: The requirements for Zero Egress is that the S3/ECR/EC2 endpoints resolve to a private IP. If we cannot do that within network verifier, that's going to not fully verify the VPC is ready for zero egress.

**Comment by bmeng** [comment]
local testing result

Testing zero egress with zero egress network:
```
$ ./osd-network-verifier egress --security-group-ids sg-0784d928ccd75965d --region us-west-2 --subnet-id subnet-0d19c770292887e50 --platform aws-hcp-zeroegress
Using region: us-west-2
Failed to get egress list from GitHub, falling back to local list: GET https://api.github.com/repos/openshift/osd-network-verifier/contents//pkg/data/egress_lists/aws-hcp-zeroegress.yaml: 404 Not Found []
Created instance with ID: i-041121525f1be8a0c
Deleting instance with ID: i-041121525f1be8a0c
Summary:
All tests passed!
Success
```

Testing zero egress with normal network:
```
$ ./osd-network-verifier egress --security-group-ids sg-4f5f0b69 --region us-west-2 --subnet-id subnet-8f8f5ec5 --platform aws-hcp-zeroegress
Using region: us-west-2
Failed to get egress list from GitHub, falling back to local list: GET https://api.github.com/repos/openshift/osd-network-verifier/contents//pkg/data/egress_lists/aws-hcp-zeroegress.yaml: 404 Not Found []
Created instance with ID: i-026830193cf71b7c0
Deleting instance with ID: i-026830193cf71b7c0
Summary:
printing out failures:
 - egressURL error: https://kms.us-west-2.amazonaws.com:443 (The endpoint is non private)
 - egressURL error: https://sts.us-west-2.amazonaws.com:443 (The endpoint is non private)
 - egressURL error: https://api.ecr.us-west-2.amazonaws.com:443 (The endpoint is non private)
 - egressURL error: https://whatever.dkr.ecr.us-west-2.amazonaws.com:443 (The endpoint is non private)

printing out exceptions preventing the verifier from running the specific test:
printing out errors faced during the execution:
Failure!
```

Testing normal hcp with normal network
```
$ ./osd-network-verifier egress --security-group-ids sg-4f5f0b69 --region us-west-2 --subnet-id subnet-8f8f5ec5 --platform aws-hcp
Using region: us-west-2
Using egress URL list from https://api.github.com/repos/openshift/osd-network-verifier/contents/pkg/data/egress_lists/aws-hcp.yaml?ref=main at SHA a48a0844430b376fcc1d229c97e9d8b527706b60
Created instance with ID: i-00423787ff2a0dc7a
Deleting instance with ID: i-00423787ff2a0dc7a
Summary:
All tests passed!
Success
```

cc @dustman9000

File: pkg/verifier/aws/aws_verifier_test.go
Comment (dustman9000) [inline pkg/verifier/aws/aws_verifier_test.go:75]: Can you also add a test for `ensurePrive = true`?

Comment (bmeng) [reply pkg/verifier/aws/aws_verifier_test.go:75]: can do, let me know if the implementation itself is good. If everything is good, I can add tests accordingly.

File: pkg/data/egress_lists/aws-hcp-zeroegress.yaml
Comment (dustman9000) [inline pkg/data/egress_lists/aws-hcp-zeroegress.yaml:5]:
```suggestion
- host: example.dkr.ecr.${AWS_REGION}.amazonaws.com
```

File: pkg/verifier/aws/aws_verifier_test.go
Comment (dustman9000) [reply pkg/verifier/aws/aws_verifier_test.go:75]: implementation looks good to me

File: pkg/probes/curl/curl_json.go
Comment (abyrne55) [inline pkg/probes/curl/curl_json.go:209]: Could we add some comments here and in this function's docstring explaining the usecase for `ensurePrivate`?

File: pkg/verifier/aws/entry_point.go
Comment (abyrne55) [inline pkg/verifier/aws/entry_point.go:249]: Please also add comments here

**Review by abyrne55 — CHANGES_REQUESTED** [review]
Code looks good, but I haven't tested it yet. While I arrange some time for that, I'm requesting some code comments be added explaining why we need the "is private" checks

**Comment by lnguyen1401** [comment]
I have tested the PR 
- Set up rosa Zero egress with https://github.com/dustman9000/rosa-hcp-zero-egress-terraform
- Run the test ./osd-network-verifier egress --subnet-id subnet-05f2eada7614b1b1b --profile linnguye --region us-west-2 --platform aws-hcp-zeroegress --debug

Result: 
@NV@{"content_type":"text/plain; charset=utf-8","errormsg":null,"exitcode":0,"filename_effective":null,"ftp_entry_path":null,"http_code":401,"http_connect":000,"http_version":"1.1","local_ip":"10.0.2.25","local_port":36370,"method":"HEAD","num_connects":1,"num_headers":6,"num_redirects":0,"proxy_ssl_verify_result":0,"redirect_url":null,"referer":null,"remote_ip":"10.0.2.74","remote_port":443,"response_code":401,"scheme":"HTTPS","size_download":0,"size_header":293,"size_request":104,"size_upload":0,"speed_download":0,"speed_upload":0,"ssl_verify_result":0,"time_appconnect":0.210874,"time_connect":0.092782,"time_namelookup":0.003309,"time_pretransfer":0.210905,"time_redirect":0.000000,"time_starttransfer":0.213725,"time_total":0.213769,"url":"https://example.dkr.ecr.us-west-2.amazonaws.com:443","url_effective":"https://example.dkr.ecr.us-west-2.amazonaws.com:443/","urlnum":1,"curl_version":"libcurl/7.76.1 OpenSSL/3.0.7 zlib/1.2.11 brotli/1.0.9 libidn2/2.3.0 libpsl/0.21.1 (+libidn2/2[2024-10-01T14:01:55.276978].3.0) libssh/0.10.4/openssl/zlib nghttp2/1.43.0"}
@NV@{"content_type":null,"errormsg":null,"exitcode":0,"filename_effective":null,"ftp_entry_path":null,"http_code":302,"http_connect":000,"http_version":"1.1","local_ip":"10.0.2.25","local_port":45302,"method":"HEAD","num_connects":1,"num_headers":4,"num_redirects":0,"proxy_ssl_verify_result":0,"redirect_url":"https://aws.amazon.com/iam","referer":null,"remote_ip":"10.0.2.246","remote_port":443,"response_code":302,"scheme":"HTTPS","size_download":0,"size_header":181,"size_request":92,"size_upload":0,"speed_download":0,"speed_upload":0,"ssl_verify_result":0,"time_appconnect":0.214308,"time_connect":0.010689,"time_namelookup":0.003802,"time_pretransfer":0.214337,"time_redirect":0.000000,"time_starttransfer":0.242830,"time_total":0.242876,"url":"https://sts.us-west-2.amazonaws.com:443","url_effective":"https://sts.us-west-2.amazonaws.com:443/","urlnum":0,"curl_version":"libcurl/7.76.1 OpenSSL/3.0.7 zlib/1.2.11 brotli/1.0.9 libidn2/2.3.0 libpsl/0.21.1 (+libidn2/2.3.0) libssh/0.10.4/ope[2024-10-01T14:01:55.277039]nssl/zlib nghttp2/1.43.0"}
@NV@{"content_type":null,"errormsg":null,"exitcode":0,"filename_effective":null,"ftp_entry_path":null,"http_code":404,"http_connect":000,"http_version":"1.1","local_ip":"10.0.2.25","local_port":40808,"method":"HEAD","num_connects":1,"num_headers":3,"num_redirects":0,"proxy_ssl_verify_result":0,"redirect_url":null,"referer":null,"remote_ip":"10.0.2.188","remote_port":443,"response_code":404,"scheme":"HTTPS","size_download":0,"size_header":143,"size_request":96,"size_upload":0,"speed_download":0,"speed_upload":0,"ssl_verify_result":0,"time_appconnect":0.213204,"time_connect":0.151844,"time_namelookup":0.003283,"time_pretransfer":0.213240,"time_redirect":0.000000,"time_starttransfer":0.242175,"time_total":0.242187,"url":"https://api.ecr.us-west-2.amazonaws.com:443","url_effective":"https://api.ecr.us-west-2.amazonaws.com:443/","urlnum":2,"curl_version":"libcurl/7.76.1 OpenSSL/3.0.7 zlib/1.2.11 brotli/1.0.9 libidn2/2.3.0 libpsl/0.21.1 (+libidn2/2.3.0) libssh/0.10.4/openssl/zlib nghttp[2024-10-01T14:01:55.277125]2/1.43.0"}
---
 - &{ContentType:text/plain; charset=utf-8 ErrorMsg: ExitCode:0 FilenameEffective: FTPEntryPath: HTTPCode:401 HTTPConnect:0 HTTPVersion:1.1 LocalIP:10.0.2.25 LocalPort:36370 Method:HEAD NumConnects:1 NumHeaders:6 NumRedirects:0 ProxySSLVerifyResult:0 RedirectURL: Referer: RemoteIP:10.0.2.74 RemotePort:443 ResponseCode:401 Scheme:HTTPS SizeDownload:0 SizeHeader:293 SizeRequest:104 SizeUpload:0 SpeedDownload:0 SpeedUpload:0 SSLVerifyResult:0 TimeAppConnect:0.210874 TimeConnect:0.092782 TimeNameLookup:0.003309 TimePreTransfer:0.210905 TimeRedirect:0 TimeStartTransfer:0.213725 TimeTotal:0.213769 URL:https://example.dkr.ecr.us-west-2.amazonaws.com:443 URLEffective:https://example.dkr.ecr.us-west-2.amazonaws.com:443/ URLNum:1 CurlVersion:libcurl/7.76.1 OpenSSL/3.0.7 zlib/1.2.11 brotli/1.0.9 libidn2/2.3.0 libpsl/0.21.1 (+libidn2/2.3.0) libssh/0.10.4/openssl/zlib nghttp2/1.43.0}

 - &{ContentType: ErrorMsg: ExitCode:0 FilenameEffective: FTPEntryPath: HTTPCode:302 HTTPConnect:0 HTTPVersion:1.1 LocalIP:10.0.2.25 LocalPort:45302 Method:HEAD NumConnects:1 NumHeaders:4 NumRedirects:0 ProxySSLVerifyResult:0 RedirectURL:https://aws.amazon.com/iam Referer: RemoteIP:10.0.2.246 RemotePort:443 ResponseCode:302 Scheme:HTTPS SizeDownload:0 SizeHeader:181 SizeRequest:92 SizeUpload:0 SpeedDownload:0 SpeedUpload:0 SSLVerifyResult:0 TimeAppConnect:0.214308 TimeConnect:0.010689 TimeNameLookup:0.003802 TimePreTransfer:0.214337 TimeRedirect:0 TimeStartTransfer:0.24283 TimeTotal:0.242876 URL:https://sts.us-west-2.amazonaws.com:443 URLEffective:https://sts.us-west-2.amazonaws.com:443/ URLNum:0 CurlVersion:libcurl/7.76.1 OpenSSL/3.0.7 zlib/1.2.11 brotli/1.0.9 libidn2/2.3.0 libpsl/0.21.1 (+libidn2/2.3.0) libssh/0.10.4/openssl/zlib nghttp2/1.43.0}

 - &{ContentType: ErrorMsg: ExitCode:0 FilenameEffective: FTPEntryPath: HTTPCode:404 HTTPConnect:0 HTTPVersion:1.1 LocalIP:10.0.2.25 LocalPort:40808 Method:HEAD NumConnects:1 NumHeaders:3 NumRedirects:0 ProxySSLVerifyResult:0 RedirectURL: Referer: RemoteIP:10.0.2.188 RemotePort:443 ResponseCode:404 Scheme:HTTPS SizeDownload:0 SizeHeader:143 SizeRequest:96 SizeUpload:0 SpeedDownload:0 SpeedUpload:0 SSLVerifyResult:0 TimeAppConnect:0.213204 TimeConnect:0.151844 TimeNameLookup:0.003283 TimePreTransfer:0.21324 TimeRedirect:0 TimeStartTransfer:0.242175 TimeTotal:0.242187 URL:https://api.ecr.us-west-2.amazonaws.com:443 URLEffective:https://api.ecr.us-west-2.amazonaws.com:443/ URLNum:2 CurlVersion:libcurl/7.76.1 OpenSSL/3.0.7 zlib/1.2.11 brotli/1.0.9 libidn2/2.3.0 libpsl/0.21.1 (+libidn2/2.3.0) libssh/0.10.4/openssl/zlib nghttp2/1.43.0}


All tests passed!
Success


The functionality looks good to me

**Comment by abyrne55** [comment]
Tested locally and confirmed no regressions. We'll handle improving docs/comments in a separate PR. Nice work @bmeng!

/lgtm
/approve

**Pull request merged**
